[CU-f94vpz] Add possibility to delete tournament

PrzemekRychter opened the pull request:
- Added posibility to delete tournament (only) by owner
- Resolved id on database layer so everything we pass to and get from repo should be populated (tournament from CreateTournament was not pupulated)
- resolved ids from our sharedTypes to _id on db level (simplicity in our types)
- with resolving _ids to id, I corrected objects returned from db, in previous version there were no needed fields in these objects
- some small changes in update score bus.log (the final score was not set if the reported results were consistent)

PrzemekRychter committed:
added posibility for tournament removal

PrzemekRychter committed:
Delete tournament

PrzemekRychter committed:
removal of the tournament checked by postman - ok

PrzemekRychter committed:
removing console.logs

PrzemekRychter committed:
added authorization of deletion of tournament

PrzemekRychter committed:
Added some validation at repositories,
decoupling db infra in every repo method

PrzemekRychter committed:
c

Zummek commented:
Task linked: [CU-f94vpz Add possibility to delete tournament](https://app.clickup.com/t/f94vpz)

PrzemekRychter committed:
added transaction session to deletion of tournament

PrzemekRychter committed:
added check in t. creation if teams were created

PrzemekRychter committed:
_id , __v.. etc resolved by repositories,
apart from repositories we can use ids now not _ids
everything chacked by postman
createTournament from Repo was returning not populated tournament now its populated

Zummek commented on server/concepts/Match.ts:
```diff
@@ -61,11 +61,11 @@ export default class Match implements MatchWithoutMS {
-    const rawMatch = await TournamentRepository.getMatchById(matchId);
+    const matchWithoutMS = await TournamentRepository.getMatchById(matchId);
 
-    if (!rawMatch) throw new AppError('Match does not exits', 404);
+    if (!matchWithoutMS) throw new AppError('Match with provided id, does not exist in any tournament', 400);
 
-    const match = new Match(rawMatch);
+    const match = new Match(matchWithoutMS);
```
imo these changes are redundant. But ok, but AppError should return 404, not 400. Previous message was short and sufficient. In changes on main branch we should return message like `Match with provided id, does not exist in the indicated tournament` or `Match does not exist in indicated tournament`

PrzemekRychter replied:
changed

Zummek commented on server/concepts/Tournament.ts:
```diff
@@ -32,8 +32,8 @@ export default class Tournament implements TournamentWihtoutMS {
 
   public static async create(data: TournamentApi.Create, ownerId: string) {
     const teams = await TeamRepository.createMany(data.teams);
+    if (teams.length !== data.teams.length) throw new AppError('Failed to register all teams', 406);
```
wrong http status. There should be a 400 here. https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400

PrzemekRychter replied:
corrected

Zummek commented on server/concepts/Tournament.ts:
```diff
@@ -42,6 +42,14 @@ export default class Tournament implements TournamentWihtoutMS {
+  public static async delete(tournamentId: string, currentUserId : string) {
+    const tournament = await TournamentRepository.getById(tournamentId);
+    if (!tournament) throw new AppError('Tournament with such ID does not exist', 404);
+    if (currentUserId !== tournament.ownerId) throw new AppError('You are not an owner of given tournament', 403);
```
```suggestion
    if (currentUserId !== tournament.ownerId) throw new AppError('You are not authorized to delete given tournament', 403);
```

PrzemekRychter replied:
```suggestion
    if (currentUserId !== tournament.ownerId) throw new AppError('You are not authorized to delete given tournament. 
Only owner of specified tournament can delete it
', 403);
```

PrzemekRychter replied:
```suggestion
    if (currentUserId !== tournament.ownerId) throw new AppError('You are not an owner of given tournament. Only owner of specified tournament can delete it', 403);
```

Zummek replied:
I mean we should return a short and concise message. if we skip information about who can do this action, we won't have to change this message and makes changes on frontend in future

Zummek commented on shared/types/Tournament/index.d.ts:
```diff
@@ -18,35 +18,35 @@ export interface MatchSide {
+  id?: string;
   name: string;
-  owner: User;
+  ownerId: string;
```
I'll scream. Why did you remove owner object? I need information about owner

PrzemekRychter replied:
Because its not normal user, imagine we add alias or other fields, User is our type for tournament participant

PrzemekRychter replied:
*participants

Zummek commented on server/routes/api-v1/tournamentRouter.ts:
```diff
@@ -4,7 +4,7 @@ import tournamentController from '../../controllers/tournamentController';
-
+tournamentRouter.get('/', tournamentController.getAllTournaments);
+tournamentRouter.get('/:id', tournamentController.getTournament);
+tournamentRouter.delete('/:id', tournamentController.deleteTournament);
```
Restore old naming

PrzemekRychter replied:
i will try during rebase

Zummek commented on server/concepts/Tournament.ts:
```diff
@@ -114,9 +122,11 @@ export default class Tournament implements TournamentWihtoutMS {
     const enrichedTournaments = [];
 
-    for (let i = 0; i < tournaments.length; i++) {
+    if (!(tournaments === [])) {
```
Is it for what? When tournaments will be empty array, the loop will not execute

PrzemekRychter replied:
deleted

Zummek commented on server/controllers/tournamentController.ts:
```diff
@@ -21,12 +21,12 @@ const getAllTournaments = catchAsync(async (req, res) => {
 
   const data = await TournamentRepository.getAll(page, pageSize);
-  const enrichedTournament = await Tournament.enrichTournamentsWithMSUsers(data.tournaments, `Bearer ${req.cookies.jwt}`);
+  const enrichedTournaments : Tournament[] = await Tournament.enrichTournamentsWithMSUsers(data.tournaments, `Bearer ${req.cookies.jwt}`);
```
Is this type required?
```suggestion
  const enrichedTournaments = await Tournament.enrichTournamentsWithMSUsers(data.tournaments, `Bearer ${req.cookies.jwt}`);
```

PrzemekRychter replied:
no , deleted

Zummek commented on server/controllers/tournamentController.ts:
```diff
@@ -46,8 +46,16 @@ const getTournament = catchAsync(async (req, res) => {
+  // updateMatchOutcomes,
+  getAllTournaments,
+  getTournament,
+  deleteTournament,
```
naming

PrzemekRychter replied:
I had error during rebase, i will try to fix this

Zummek commented on server/database/models/TournamentModel.ts:
```diff
@@ -1,22 +1,25 @@
+}, { _id: false });
 
 const MatchSchema = new Schema({
+  // _id added automatically
```
It's redundant. Developer should know this

PrzemekRychter replied:
It's not, it reminds about all fields in db. I disabled _id for some subdocuments (which you can do only for subdocuments)

Zummek commented on server/database/models/TournamentModel.ts:
```diff
@@ -32,4 +35,18 @@ TournamentSchema.pre(/^find/, function (next) {
+  sideA: MatchSideDocument
+  sideB: MatchSideDocument
+
+}
```
```suggestion
export interface MatchDocument extends Omit<MatchWithoutMS, 'id' | 'sideA' | 'sideB'>, Document {
  sideA: MatchSideDocument
  sideB: MatchSideDocument
}
```
Is Linter not working?

PrzemekRychter replied:
Idk, it is.

Zummek commented on server/database/repositories/TeamRepository.ts:
```diff
@@ -1,11 +1,66 @@
+    // .. more user fields
+  };
+  return usr;
+}
```
```suggestion
function toUser(usrDb: UserDocument): UserWithoutMS {
  // const usr : UserWithoutMS = {id: usrDb._id, ...usrDb}; <- tak nie mozna bo zworcimy wszystkie rzeczy z mongo
  return {
    id: usrDb._id.toString(),
    // .. more user fields
  };
}
```
So why can't we use .lean()? When you will get a lot of properties, changes will be crazy

PrzemekRychter replied:
Probably won't add a lot of properties. It will not be crazy, since returned type is UserWithoutMS, TS will  not allow you to run without specifying all UserWithoutMS fields.

Zummek commented on server/database/repositories/TeamRepository.ts:
```diff
@@ -1,11 +1,66 @@
+function toUserDb(usr: UserWithoutMS): UserDb {
+  const usrDb : UserDb = { _id: usr.id, ...usr };
+  return usrDb;
+}
```
```suggestion
function toUserDb(usr: UserWithoutMS): UserDb {
  return { _id: usr.id, ...usr };
}
```

When you add returns type, you can return object like this. This also applies to the rest

PrzemekRychter replied:
corrected

Zummek commented on server/database/repositories/TeamRepository.ts:
```diff
@@ -1,11 +1,66 @@
+  public static createMany = async (teams: TeamWithoutMS[]) => {
+    const teamsDb = teams.map((team) => toTeamDb(team));
+
+    teamsDb.forEach(async (team) => {
+      if (team._id !== undefined) throw new AppError('New team should not contain id', 400);
+    });
```
Same as above. I think this is unnecessary checking

PrzemekRychter replied:
it is

Zummek commented on server/database/repositories/TeamRepository.ts:
```diff
@@ -1,11 +1,66 @@
-  public static create = async (team: TeamWithoutMS) => TeamModel.create(team);
+  public static create = async (team: TeamWithoutMS) => {
+    const teamDb = toTeamDb(team);
+    if (teamDb._id !== undefined) throw new AppError('New team should not contain id', 400);
```
I think this is unnecessary checking

PrzemekRychter replied:
it's necessary because you could override existing tournament with that method. id -> _id (toTeamDb)

Zummek commented on server/database/repositories/TeamRepository.ts:
```diff
@@ -1,11 +1,66 @@
+    teamsDb.forEach(async (team) => {
+      if (team._id !== undefined) throw new AppError('New team should not contain id', 400);
+    });
+    const teamDocs : TeamDocument[] = await TeamModel.insertMany(teamsDb);
```
```suggestion
    const teamDocs = await TeamModel.insertMany(teamsDb);
```
Is this type required?

PrzemekRychter replied:
no deleted

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -1,37 +1,114 @@
+  team: TeamDb,
+}
+interface MatchDb extends Omit<MatchWithoutMS, 'id' | 'sideA' | 'sideB'> {
+  _id?:string,
```
```suggestion
  _id?: string,
```

PrzemekRychter replied:
corrected

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -1,37 +1,114 @@
   public static async create(tournament: TournamentWihtoutMS) {
-    // Check if teams have ids, if teams in matches have ids
-    // and if teams in matches are on teams list
+    if (tournament.id) { throw new AppError('New tournament should not have id', 400); }
```
Same as above. I think this is unnecessary checking

PrzemekRychter replied:
it is necessary because we allow to pass tournament with id to this function

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -1,37 +1,114 @@
     });
-
-    return await TournamentModel.create(tournament);
+    const tDoc = await TournamentModel.create(toTournamentDb(tournament));
```
Use `Doc` or `Db`

PrzemekRychter replied:
Db and Doc is diffferent
by Doc i reffer to Document which is object returned from mongoose
by ...Db types i represent objects from shared types but with ids mapped to _id

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -1,37 +1,114 @@
     });
-
-    return await TournamentModel.create(tournament);
+    const tDoc = await TournamentModel.create(toTournamentDb(tournament));
+    const tour = await TournamentRepository.getById(tDoc._id);
+    if (!tour) { throw new AppError('Error while creating tournament', 400); }
+    return tour;
```
`TournamentModel.create` returns tournament, so checking / getting it from database is redundant. Database throw error when it occurs

PrzemekRychter replied:
But it was not populated. So i needed to populate somehow, i used getById method, it can returns null so i need to check that, even though it probably never will

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -1,37 +1,114 @@
+  public static async updateMatch(match: MatchWithoutMS) {
+    if (!match.id) throw new AppError('Provided match does not contain id', 400);
+    const matchDb = toMatchDb(match);
+    const tDoc: TournamentDocument | null = await TournamentModel.findOneAndUpdate(
```
```suggestion
    const tDoc = await TournamentModel.findOneAndUpdate(
```
required?

PrzemekRychter replied:
no, deleted

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -42,17 +119,35 @@ export default class TournamentRepository {
     }]);
-    const tournaments: TournamentWihtoutMS[] = await TournamentModel.find().skip(skip).limit(pageSize).lean();
-
+    const tDocs: TournamentDocument[] = await TournamentModel.find().skip(skip).limit(pageSize).exec();
+    const tournaments: TournamentWihtoutMS[] = tDocs.map((tDoc) => toTournament(tDoc));
```
```suggestion
    const tDocs = await TournamentModel.find().skip(skip).limit(pageSize).exec();
    const tournaments = tDocs.map((tDoc) => toTournament(tDoc));
```
?

PrzemekRychter replied:
Mapping TournamentDocument [] to TournamentWithoutMS ( mapping _ids to id and ommiting mongoose.Document fields)

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -1,37 +1,114 @@
-
-    return tournament;
+    ).exec();
+    if (!tDoc) throw new AppError('Attempting to update a non-existent match', 400);
```
```suggestion
    if (!tDoc) throw new AppError('Match does not exits', 404);
```
But is necessary to checking this? I wonder if we should check this before calling this method in business logic

PrzemekRychter replied:
If we decide to update it, we have to be sure that specific match exists, we can return null in method like : getMatchById, and then we can check if match is null in b.log.

PrzemekRychter replied:
What's the point of trying to update match that we even don't know if it exists

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -42,17 +119,35 @@ export default class TournamentRepository {
 
   public static async getMatchById(matchId: string) {
-    const tournament = await TournamentModel.findOne({ 'matches._id': matchId }).lean();
+    const tDoc = await TournamentModel.findOne({ 'matches._id': matchId }).exec();
+    if (!tDoc) return null;
+    const matchDoc = tDoc.matches.find((match) => match._id === matchId);
+    if (!matchDoc) return null;
+    return toMatch(matchDoc);
```
Make same spaces between lines for better readability. Same below

PrzemekRychter replied:
ok

PrzemekRychter replied:
spaces before return ?

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -42,17 +119,35 @@ export default class TournamentRepository {
+    const tDoc = await TournamentModel.findById(id).exec();
+    if (!tDoc) throw new AppError('Can not delete tournament which does not exist', 400);
+    const session = await TournamentModel.startSession();
+    await session.withTransaction(async () => {
+      tDoc.teams.forEach(async (team) => {
+        await TeamModel.findByIdAndDelete(team._id).exec();
+      });
+      await tDoc.deleteOne();
+    });
+    session.endSession();
```
I'm not sure but you should add `session` param to methods param

PrzemekRychter replied:
idk, it's working, i had no idea how to test it

PrzemekRychter committed:
Update server/concepts/Tournament.ts

PrzemekRychter committed:
cor

PrzemekRychter committed:
Corrections after CR

PrzemekRychter committed:
corrected after cr

PrzemekRychter committed:
m

PrzemekRychter committed:
- updateMatchScore logic works if scores aren't negative
-if scores was reported and they are the same we should not only set isFinished to true, but also final score

Zummek commented on server/concepts/Tournament.ts:
```diff
@@ -57,10 +70,12 @@ export default class Tournament implements TournamentWihtoutMS {
     matchId: string,
     currentUserId: string,
   ) {
+    if (data.teamA < 0 || data.teamB < 0) throw new AppError('Score can\'t be negative', 404);
```
Hmm score can't be negative and must be integer
Don't use shortcuts, I mean replace `can't` to `can not`

PrzemekRychter replied:
Ok

Zummek commented on server/concepts/Tournament.ts:
```diff
@@ -81,19 +96,24 @@ export default class Tournament implements TournamentWihtoutMS {
-
-    await TournamentRepository.updateMatch(matchId, match);
+    // If both reported scores are the same, && both teams reported scores
+    if (match.score.reportedByA === match.score.reportedByB && match.score.reportedByA.a !== -1) {
+      match.score.final = {
+        a: match.score.reportedByA.a,
+        b: match.score.reportedByA.b,
+      };
+      match.isFinished = true;
+    }
```
It's not good. Bcz when owner want to overwrite scores, this cod/if will overwrite again

PrzemekRychter replied:
It will not override in any case

PrzemekRychter replied:
Only one case when this code will override score updated by owner is when owner will try to update score which was already updated by both teams, and is the same. Do we want to allow owner to override score after it is finished?

PrzemekRychter replied:
I thought that owner would have only one chance to set final score, now owner can always set scores.

Zummek reviewed (approved):
Remove redundant comments, look to comments and will be ok. @eastemployeer it's your turn to CR

eastemployeer commented on server/database/models/TournamentModel.ts:
```diff
@@ -1,18 +1,20 @@
 import mongoose, { Schema, Document } from 'mongoose';
-import { TournamentWihtoutMS } from '../../../shared/types/Tournament';
+import { MatchWithoutMS, TournamentWihtoutMS } from '../../../shared/types/Tournament';
```
Name of the interface is wrong `TournamentWithoutMS`

PrzemekRychter replied:
?, corrected

eastemployeer commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -1,37 +1,89 @@
+  public static async updateMatch(match: MatchWithoutMS) {
+    if (!match.id) throw new AppError('Provided match does not contain id', 400);
+    const matchDb = toMatchDb(match);
+    const tDoc = await TournamentModel.findOneAndUpdate(
```
Bad variable name. Means nothing

PrzemekRychter replied:
corrected

eastemployeer commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -1,37 +1,89 @@
-    ).lean();
-
-    return tournament;
+    ).exec();
```
Why exec()?

PrzemekRychter replied:
https://mongoosejs.com/docs/promises.html#should-you-use-exec-with-await?

eastemployeer commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -42,17 +94,35 @@ export default class TournamentRepository {
-  public static getById =
-  async (id: string): Promise<TournamentWihtoutMS | null> => TournamentModel.findById(id).lean();
+  public static getById = async (id: string) => {
+    const tDoc: TournamentDocument | null = await TournamentModel.findById(id).exec();
```
Once again. Why exec()

eastemployeer commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -42,17 +94,35 @@ export default class TournamentRepository {
 
   public static async getMatchById(matchId: string) {
-    const tournament = await TournamentModel.findOne({ 'matches._id': matchId }).lean();
+    const tDoc = await TournamentModel.findOne({ 'matches._id': matchId }).exec();
```
Why exec(). Delete it everywhere

PrzemekRychter committed:
Finished match can not be updated

PrzemekRychter committed:
after Kamil CR

PrzemekRychter committed:
+

PrzemekRychter committed:
If reported scores are the same final score is set the same as them
